### PR TITLE
Alias traceable_allocator to std::allocator when building without GC

### PR DESCRIPTION
**Pull request by edolstra**
# Motivation

This gets rid of a bunch of #ifdefs and  avoids build errors like the one mentioned in https://github.com/NixOS/nix/pull/11548#pullrequestreview-2316345551.

Depends on #11548.

# Context

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: src/libexpr/value.hh [hunk 1]
```diff
@@ -12,7 +12,11 @@
 
 #if HAVE_BOEHMGC
 #include <gc/gc_allocator.h>
+#else
+template<typename T>
+using traceable_allocator = std::allocator<T>;
```
Comment (roberth) [inline src/libexpr/value.hh:17]: For consistency, also do `gc_allocator`?

File: src/libexpr/gc-small-vector.hh [hunk 2]
```diff
@@ -2,28 +2,15 @@
...
 #include <boost/container/small_vector.hpp>
 
-#if HAVE_BOEHMGC
-
-#include <gc/gc.h>
-#include <gc/gc_cpp.h>
-#include <gc/gc_allocator.h>
-
-#endif
+#include "value.hh"
```
Comment (roberth) [inline src/libexpr/gc-small-vector.hh:5]: I'd like for this header to be more of a "leaf" considering its narrow responsibility, so importing `value.hh` here is not great. How about we put it in `eval-gc.hh`?

**Review by roberth — COMMENTED** [review]
This is much nicer

File: src/libexpr/eval-gc.hh [hunk 3]
```diff
@@ -3,6 +3,31 @@
...
+
+template<typename T>
+using traceable_allocator = std::allocator<T>;
+
+template<typename T>
+using gc_allocator = std::allocator<T>;
+
+#define GC_MALLOC_ATOMIC std::malloc
+#define GC_STRDUP std::strdup
+struct gc { };
```
Comment (Mic92) [inline src/libexpr/eval-gc.hh:27]:
```suggestion
#  define GC_MALLOC_ATOMIC std::malloc
#  define GC_STRDUP std::strdup
struct gc { };
```

**Review by Mic92 — APPROVED** [review]

**Pull request merged**